vshard/replicaset.lua

```diff
@@ -75,6 +75,141 @@ local fiber_cond_wait = util.fiber_cond_wait
+    vconn.future = conn:call('vshard.storage._call', {'info'}, opts)
+    vconn.future_cond:broadcast()
+end
+
+--
+-- Check, that future is ready, and its result is expected.
+-- The function doesn't yield.
+-- @retval true; The correct response is received.
+-- @retval nil, ...; Response is not received or validation error happened.
+--
+local function conn_vconnect_check(conn)
+    local vconn = conn.vconnect
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one. It's also nil, when
+    -- all checks were already done.
+    if not vconn then
+        return true
+    end
+    -- Nothing to do, but wait in such case.
+    if not vconn.future or not vconn.future:is_ready() then
+        return nil, lerror.vshard(lerror.code.VHANDSHAKE_NOT_COMPLETE,
+                                  conn.replica.id)
+    end
+    -- Critical errors. Connection should be closed after these ones.
+    local result, err = vconn.future:result()
+    if not result then
+        -- Failed to get response. E.g. access error.
+        return nil, lerror.make(err)
+    end
+    if vconn.is_named and result[1].name ~= conn.replica.name then
+        return nil, lerror.vshard(lerror.code.INSTANCE_NAME_MISMATCH,
+                                  conn.replica.name, result[1].name)
+    end
+    -- Don't validate until reconnect happens.
+    conn.vconnect = nil
+    return true
+end
+
+local function conn_vconnect_check_and_close(conn)
+    local ok, err = conn_vconnect_check(conn)
+    if not ok and err.code ~= lerror.code.VHANDSHAKE_NOT_COMPLETE then
+        conn:close()
+    end
+    return ok, err
+end
+
+--
+-- Wait until the future object is ready. Returns remaining timeout.
+-- @retval timeout; Future boject is ready.
+-- @retval nil, err; Timeout passed.
+--
+local function conn_vconnect_wait(conn, timeout)
+    local vconn = conn.vconnect
+    -- Fast path. In most cases no validation should be done at all.
+    -- conn.vconnect may be nil, if connection was created on old version
+    -- and the storage was reloaded to a new one. It's also nil, when
+    -- all checks were already done.
+    if not vconn or (vconn.future and vconn.future:is_ready()) then
+        return timeout
+    end
+    local deadline = fiber_clock() + timeout
+    -- Wait for connection to be established.
+    if not vconn.future and
+       not fiber_cond_wait(vconn.future_cond, timeout) then
+        return nil, lerror.timeout()
+    end
+    timeout = deadline - fiber_clock()
+    -- Wait for async call to return.
+    local res, err = util.future_wait(vconn.future, timeout)
+    if res == nil then
+        -- Either timeout error or any other. If it's not a timeout error,
+        -- then conn must be recteated, handshake must be retried.
+        return nil, lerror.make(err)
+    end
+    return deadline - fiber_clock()
+end
+
+local function conn_vconnect_wait_and_close(conn, timeout)
```

Comment:
Given how it works, it is rather `wait_or_close`, not `and`. Same for `check_and_close`.

Reply:
Renamed